manager.go

```diff
@@ -1,118 +1,141 @@
+	duringDrainHooks []func()
 }
 
-func (m *manager) start() {
-	m.Add(1)
-	m.loadWorkers()
-	go m.manage()
+func NewManager(options Options) (*Manager, error) {
+	options, err := processOptions(options)
+	if err != nil {
+		return nil, err
+	}
+	return &Manager{
+		uuid: uuid.New(),
+		opts: options,
+	}, nil
 }
 
-func (m *manager) prepare() {
-	if !m.fetch.Closed() {
-		m.fetch.Close()
+func (m *Manager) AddWorker(queue string, concurrency int, job JobFunc, mids ...MiddlewareFunc) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	middlewareQueueName := m.opts.Namespace + queue
+	if len(mids) == 0 {
+		job = DefaultMiddlewares().build(middlewareQueueName, m, job)
+	} else {
+		job = NewMiddlewares(mids...).build(middlewareQueueName, m, job)
 	}
+	m.workers = append(m.workers, newWorker(queue, concurrency, job))
 }
 
-func (m *manager) quit() {
-	Logger.Println("quitting queue", m.queueName(), "(waiting for", m.processing(), "/", len(m.workers), "workers).")
-	m.prepare()
+func (m *Manager) AddBeforeStartHooks(hooks ...func()) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.beforeStartHooks = append(m.beforeStartHooks, hooks...)
+}
 
-	m.workersM.Lock()
-	for _, worker := range m.workers {
-		worker.quit()
-	}
-	m.workersM.Unlock()
+func (m *Manager) AddDuringDrainHooks(hooks ...func()) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.duringDrainHooks = append(m.duringDrainHooks, hooks...)
+}
 
-	m.stop <- true
-	<-m.exit
+// Run starts all workers under this Manager and blocks until they exit.
+func (m *Manager) Run() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.running {
+		return // Can't start if we're already running!
+	}
+	m.running = true
 
-	m.reset()
+	for _, h := range m.beforeStartHooks {
+		h()
+	}
 
-	m.Done()
-}
+	globalStatsServer.registerManager(m)
 
-func (m *manager) manage() {
-	Logger.Println("processing queue", m.queueName(), "with", m.concurrency, "workers.")
+	var wg sync.WaitGroup
 
-	go m.fetch.Fetch()
+	wg.Add(1)
+	m.signal = make(chan os.Signal, 1)
+	go func() {
+		m.handleSignals()
+		wg.Done()
+	}()
 
-	for {
-		select {
-		case message := <-m.confirm:
-			m.fetch.Acknowledge(message)
-		case <-m.stop:
-			m.exit <- true
-			break
-		}
+	wg.Add(len(m.workers))
+	for i := range m.workers {
+		w := m.workers[i]
+		go func() {
+			w.start(newSimpleFetcher(w.queue, m.opts))
+			wg.Done()
+		}()
 	}
+	m.schedule = newScheduledWorker(m.opts)
+
+	wg.Add(1)
+	go func() {
+		m.schedule.run()
+		wg.Done()
+	}()
+
+	// Release the lock so that Stop can acquire it
+	m.lock.Unlock()
+	wg.Wait()
+	// Regain the lock
+	m.lock.Lock()
+	globalStatsServer.deregisterManager(m)
+	m.running = false
 }
 
-func (m *manager) loadWorkers() {
-	m.workersM.Lock()
-	for i := 0; i < m.concurrency; i++ {
-		m.workers[i] = newWorker(m)
-		m.workers[i].start()
+// Stop all workers under this Manager and returns immediately.
+func (m *Manager) Stop() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if !m.running {
+		return // We're not running.
```

Comment:
Small nit. Do we need this comment? `if !m.running { return }` seems self-explanatory. Plus there's a comment for the function itself.